pygpu/ufuncs.py

```diff
@@ -0,0 +1,1330 @@
+
+    axis, out, keepdims :
+        Arguments as in NumPy reductions. See e.g. `numpy.amax`.
+
+    Returns
+    -------
+    reduced : `pygpu.gpuarray.GpuArray` or scalar
+        If not all axes are reduced over or ``out`` was given, the result is
+        an array, a reference to ``out`` if provided. Otherwise, i.e. for
+        reductions along all axes without ``out`` parameter, the result
+        is a scalar.
+    """
+    a = _prepare_array_for_reduction(a, out)
+
+    axes = axis if axis is not None else tuple(range(a.ndim))
+    if out is not None:
+        out_arr = out.reshape(reduce_dims(a.shape, axes))
+    else:
+        out_arr = out
+
+    oper = '(a {} b) ? a : b'.format(cmp)
+    r = reduce1(a, op=None, neutral=neutral, out_type=a.dtype, axis=axis,
+                out=out_arr, oper=oper)
+    if keepdims:
+        axes = axis if axis is not None else tuple(range(a.ndim))
+        newshape = restore_reduced_dims(r.shape, axes)
+        r = r.reshape(newshape)
+    if not r.shape and out is None:
+        return numpy.asarray(r).reshape([1])[0]
+    elif out is not None:
+        return out
+    else:
+        return r
+
+
+def amin(a, axis=None, out=None, keepdims=False):
+    """Return the minimum of an array or minimum along an axis.
+
+    See Also
+    --------
+    numpy.amin
+    """
+    a = _prepare_array_for_reduction(a, out)
+    if a.dtype == numpy.bool:
+        neutral = 1
+    elif numpy.issubsctype(a.dtype, numpy.integer):
+        neutral = numpy.iinfo(a.dtype).max
+    elif numpy.issubsctype(a.dtype, numpy.floating):
+        neutral = 'INFINITY'
+    elif numpy.issubsctype(a.dtype, numpy.complexfloating):
+        raise ValueError('array dtype {!r} not comparable'
+                         ''.format(a.dtype.name))
+    else:
+        raise ValueError('array dtype {!r} not supported'
+                         ''.format(a.dtype.name))
+    return reduce_with_cmp(a, '<', neutral, axis, out, keepdims)
+
+
+def amax(a, axis=None, out=None, keepdims=False):
+    """Return the maximum of an array or minimum along an axis.
+
+    See Also
+    --------
+    numpy.amax
+    """
+    a = _prepare_array_for_reduction(a, out)
+    if a.dtype == numpy.bool:
+        neutral = 0
+    elif numpy.issubsctype(a.dtype, numpy.integer):
+        neutral = numpy.iinfo(a.dtype).min
+    elif numpy.issubsctype(a.dtype, numpy.floating):
+        neutral = '-INFINITY'
+    elif numpy.issubsctype(a.dtype, numpy.complexfloating):
+        raise ValueError('array dtype {!r} not comparable'
+                         ''.format(a.dtype.name))
+    else:
+        raise ValueError('array dtype {!r} not supported'
+                         ''.format(a.dtype.name))
+    return reduce_with_cmp(a, '>', neutral, axis, out, keepdims)
+
+
+# --- Elementwise ufuncs --- #
+
+
+# This dictionary is derived from Numpy's C99_FUNCS list, see
+# https://github.com/numpy/numpy/search?q=C99_FUNCS
```

Comment:
Take care with this because the code you are generating is not C99, but CUDA or OpenCL.  I'm pretty sure at least some of these will not be available, like `cbrt`.